TorXakis interface

keremispirli opened the pull request:
Wrapping TorXakis functionality with a library to provide programming interface.
Build a web server that uses this library.

This PR is for review and QA support during development; do not merge yet.

Commit:
First definition of the `TorXakis` library, which features the functions needed for stepping through a model.

tretmans committed:
Added docs in Txs-Lib

Commit:
The first version of the `load` function that parses the string passed as parameter.

Commit:
First working version of the `load` function, plus instructions in the readme
on how to use it from the GHCi prompt.

Commit:
The first working version of the `load` function.
A new package `txs-lenses` that defined the first TorXakis lenses.

Commit:
A message handler that puts messages input a transactional (closable) queue `TMQueue`.
A `getNextMsg` function, that allows to retrieve the messages one step at a time.

Commit:
The first version of the `step` function.

Commit:
A Limitations section that explains the broken-pipe problem one will get when
pressing `Ctrl+c` at the GHCi REPL.

Commit:
A simplified version of the library, where a new `runIOC` function was defined.
An aynchronous version of `step`.

Commit:
change the actions reported by the stepper to pass the entire action instead of
a string representation.

created an example file.

Commit:
Incorporate the first test for the library, through an example.
Make the `runIOC` function, and hence `step` thread safe.
Add support for getting the verdicts (if any).

Commit:
Update the README to show how to use the library at the REPL.

Commit:
Incorporated `hlint` warnings.

Commit:
Merged develop

Commit:
txs-webserver WIP [skip ci]

Commit:
txs-webserver WIP [skip ci]

Commit:
Add a `/session/:sessionId/model` endpoint that prints the contents of the form that was uploaded.

Commit:
Split web server into modules

Commit:
Migrated to hpack[skip ci]

Commit:
Implemented generating unique session ids[skip ci]

Commit:
Disabled /users endpoint; only needed for tests

Commit:
Implemented loading a file into TorXakis session[skip ci]

Commit:
Forced Lib to parse input file during load[skip ci]

Commit:
Implemented starting stepper. Skipping empty files now.[skip ci]

Commit:
Added endpoint for taking n steps

Commit:
Enabled tests[skip ci]

Commit:
Added test to upload files

Commit:
Refactored UploadEP to return informative messages. Turns out tests fail; will rewrite with wreq.

Commit:
Rewriting with wreq - WIP [skip ci]

Commit:
Tests work. Remember to run the server separately.

Commit:
Added test case for starting stepper and taking 3 steps

Commit:
Now a server is launched before tests and then killed [skip ci]

Commit:
WIP[skip ci]

Commit:
Streaming messages works but can't be parsed if sent as JSON[skip ci]

Commit:
Make `txs-webserver` compile on linux.
Modify the "messages" endpoint to return JSON.

Commit:
Add SSE support.

Commit:
WIP - making sse EP respond[skip ci]

Commit:
- Correct the `waitForMessageQueue` function.

- Add termination to the SSE endpoint when the verdict is reached and the
  session message queue is empty.

Commit:
Ready to implement a demo SPA

Commit:
Fixed absolute paths in tests [skip ci]

Commit:
test page - wip

Commit:
Merged develop

Commit:
Test page works [skip ci]

Removed event type; Firefox doesn't support it, events have to start
 with "data:"
Run server in TorXakis folder and navigate to
 localhost:8080/test/test.html
BUG: We should embed this file into txs-webserver binary
Not so beautiful page, just a demo
BUG: Multiple clients mess up showing messages because of TQueue usage

Commit:
Updated readme with instructions for demo [skip ci]

Commit:
Fixed the only hlint hint. Let's see what codeclimate says.

Commit:
Fixed copyright notices

Commit:
Moved TorXakis.Session to TorXakis.Lib.Session

Commit:
Removed DataKinds and TypeOperators extensions as they are already in package.yaml

Commit:
Removed unnecessary comments

Commit:
Removed unused dependencies[skip ci]

pjljvandelaar commented on stack.yaml:
```diff
@@ -58,12 +58,22 @@ packages:
+- servant-swagger-1.1.5
+- http-types-0.12.1
+- text-1.2.3.0
 
```
why are these extra-deps needed?
servant-0.13 is in lts 11.1, like text-1.2.3.0.....

keremispirli replied:
I'll check if it works without them.

dnadales replied:
These imports were needed to be able to use `hoistServer`. Now we're just using the `Handler` provided by servant. We shoudn't need these anymore, so I'd guess they can be safely removed.

pjljvandelaar commented on sys/txs-lib/license.txt:
```diff
@@ -0,0 +1,29 @@
+BSD 3-Clause License
+
+Copyright (c) 2017, TorXakis
```
Do we need to update to 2018?
Or change license to something like  2017 - 2018 
or just 2017 -

keremispirli replied:
Do we need this licence file at all? We already have ours in the [root](https://github.com/TorXakis/TorXakis/blob/feat/txs-interface/LICENSE). It also says 2017 though...

pjljvandelaar reviewed (commented):
2 minor comments

dnadales commented on sys/txs-lenses/README.md:
```diff
@@ -0,0 +1,14 @@
+
```
I think this README is not relevant here.

pjljvandelaar committed:
Added TorXakis With Echo model

pjljvandelaar committed:
Added Info cmd + response

pjljvandelaar committed:
Model of TorXakis ready.

pjljvandelaar committed:
adding purpose

pjljvandelaar committed:
removed wrong file

pjljvandelaar committed:
Model with parallel test purpose

My proposal: feedback is more than welcome!

use
param param_IncrementChoice_IntPower 1
param param_IncrementChoice_IntRange 128
for best random effects ;-)

Can't use PURPDEF since `stepper` only support a model as argument.

pjljvandelaar committed:
minor corrections

Commit:
Fixed documentation

Commit:
Merge branch 'develop' of github.com:TorXakis/TorXakis into feat/txs-interface

Commit:
Merge branch 'feat/txs-interface' of https://github.com/TorXakis/TorXakis into feat/txs-interface

Commit:
Fix the instance of `PShow` for `Msg`.

The previous instance assumed that there was an `s` field selector for `Msg`,
which is not true anymore.

Commit:
Merge branch 'feat/txs-interface' of https://github.com/TorXakis/TorXakis into feat/txs-interface

Commit:
Implemented info in txs-lib

Commit:
Add a proposal on how to connect `TorXakis` with the SUT via HTTP.

Commit:
Added Info endpoint to txs-server[skip ci]

Commit:
Merge branch 'feat/txs-interface'[skip ci]

Commit:
Add a `tester` function to `Lib`.

Commit:
Added tester and test steps endpoints[skip ci]

Commit:
Add placeholders for the to and from world actions.

Commit:
Merge branch 'feat/txs-interface' of github.com:TorXakis/TorXakis into feat/txs-interface

Commit:
- Fix the tests of `txs-webserver`.
- Add a hanldler at `runIOC'` that puts an error originated at `TxsCore` in the
  `Torxakis.Lib` message queue.

Commit:
Add error handling so that a verdict is always produced, even in the event of a failure of the core.

Commit:
Returning bogus True to make TorXakis fail[skip ci]

Commit:
WIP [skip ci]

Commit:
Make the `txs-lib` tests pass for the `TorXakisWithEchoInfoOnly` model.

We still need to:
- [ ] Factor out common functionality from `sys/txs-lib/src/TorXakis/Lib/Examples.hs`.
- [ ] Start the `txs-webserver` and make an actual call to it.

Commit:
WIP [skip ci]

Removed some duplication from SocketWorld.hs
putToW now keeps waiting for SUT output while preparing the next Action

Commit:
Removed some todo's[skip ci]

Commit:
Removed unused files and extra-deps

Commit:
Fixed race condition in putToW [skip ci]

Commit:
Why does testPutToWReadsWorld fail?[skip ci]

Commit:
Found & fixed the problem [skip ci]

Full evaluation was causing the error to be evaluated and was not necessary
 anyway; force was enough.
Implemented putToW ActQui case

Commit:
WIP [skip ci]

Commit:
Added Sim and Sut delta times as parameters to SessionSt

Commit:
Using HTTP to test TorXakis with TorXakis[skip ci]

TorXakisWithEchoInfoOnly.txs model is used to test a TorXakis.
 The model tests INFO command.

Commit:
Improved apply func to handle Nothing from signHandler lookup

Commit:
WIP implement listening to SSE[skip ci]

Commit:
Load action is called and creates ResponseSuccess

Commit:
Refactored creation of Response Actions

Commit:
Stepper and Step functions are called - TODO:parse SSE and extract Action

Commit:
WIP[skip ci]

Commit:
Implemented extracting Msg from SSE and creating ResponseAction with it[skip ci]

Testing TorXakis with Echo works!

Commit:
Removed some useless TODOs [skip ci]

Commit:
- Add a new proposal for HTTP connection definitons.
- Make the paths in `Example.hs` platform independent.

Commit:
fixed refactoring

Commit:
Changed file upload endpoint verb to PUT

Commit:
URI parts of resources should be plural

Commit:
Updating webserver tests to reflect proper endpoints [skip ci]

Commit:
Updated Spec to use proper API endpoints; many tests fail

Commit:
Fixed response of session creation

Commit:
Implemented new model upload endpoint

Commit:
Fixed wrong file test, added case for multiple files[skip ci]

Commit:
Added test for uploading multiple model files

Commit:
Implemented new stepper endpoint

Removes streaming messages endpoint
 Now the SSE messages endpoint is the only implementation
Removes common language pragmas that are already defined in package.yaml

Commit:
Implemented counting actions in sse messages[skip ci]

Commit:
Added example to Lib to test stopping TxsCore before all steps are finished

Commit:
Bump txs-webserver version to 0.2.0.0

Commit:
Added txs-cli package[skip ci]

Commit:
Implementing CLI - WIP [skip ci]

Receive and dispatch command
Implemented background writer to be used for printing SSEs (todo)
Launch webserver before CLI
Create a new TorXakis session before starting CLI
Passing host address and session id in ReaderT Env
Implemented getting and showing TxsInfo

Commit:
Implemented uploading files via command line [skip ci]

Extracted Env

TODO:
 - Now that common functionality is visible, extract command dispatch
   implementations into their own modules.
 - Does it worth to add unit tests for the functionality up to running
   stepper? Not much logic so far...
